File: sources/Android/android-communications/library/src/main/java/com/polar/androidcommunications/api/ble/model/gatt/client/BleHtsClient.kt
package com.polar.androidcommunications.api.ble.model.gatt.client

import com.polar.androidcommunications.api.ble.BleLogger
import com.polar.androidcommunications.api.ble.BleLogger.Companion.d
import com.polar.androidcommunications.api.ble.model.gatt.BleGattBase
import com.polar.androidcommunications.api.ble.model.gatt.BleGattTxInterface
import com.polar.androidcommunications.common.ble.AtomicSet
import com.polar.androidcommunications.common.ble.ChannelUtils
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.onCompletion
import kotlinx.coroutines.flow.onStart
import java.nio.ByteBuffer
import java.util.*
import kotlin.math.pow
import kotlin.math.roundToInt

class HealthThermometer {
    companion object {
        val HTS_SERVICE: UUID = UUID.fromString("00001809-0000-1000-8000-00805f9b34fb")
        val TEMPERATURE_MEASUREMENT: UUID = UUID.fromString("00002A1C-0000-1000-8000-00805f9b34fb")
        val TEMPERATURE_TYPE: UUID = UUID.fromString("00002A1D-0000-1000-8000-00805f9b34fb")
    }
}

class BleHtsClient(txInterface: BleGattTxInterface) :
    BleGattBase(txInterface, HealthThermometer.HTS_SERVICE) {
    data class TemperatureMeasurement(
        val temperatureCelsius: Float,
        val temperatureFahrenheit: Float
    )

    companion object {
        val TAG: String = ::javaClass.name
        const val TEMP_ACCURACY: Int = 100
    }

    private val htsObserverAtomicList = AtomicSet<Channel<TemperatureMeasurement>>()

    override fun reset() {
        super.reset()
        ChannelUtils.postDisconnectedAndClearList(htsObserverAtomicList)
    }

    override fun processServiceData(
        characteristic: UUID,
        data: ByteArray,
        status: Int,
        notifying: Boolean
    ) {
        BleLogger.d(
            TAG, "processServiceData uuid=$characteristic status=$status" +
                    "notifying=$notifying len(data)=${data?.size}"
        )
        if (status == ATT_SUCCESS && data.isNotEmpty()) {
            if (characteristic == HealthThermometer.TEMPERATURE_MEASUREMENT) {
                BleLogger.d(
                    TAG,
                    "TEMPERATURE_MEASUREMENT ${data.let { BleLogger.byteArrayToHex(it) }}"
                )
                val flags = data[0].toInt() and 0xFF
                val isFahrenheit = (flags and 0x01) != 0
                val exponent = data[4].toInt()
                val mantissaBytes = byteArrayOf(0x00) + data.sliceArray(1..3).reversedArray()
                val mantissa = ByteBuffer.wrap(mantissaBytes).int
                val temperature = ((mantissa * 10.toFloat()
                    .pow(exponent) * TEMP_ACCURACY).roundToInt() / TEMP_ACCURACY.toFloat())

                val celsius =
                    if (!isFahrenheit) temperature else (temperature - 32.0f) * 5.0f / 9.0f
                val fahrenheit =
                    if (isFahrenheit) temperature else temperature * 9.0f / 5.0f + 32.0f
                ChannelUtils.emitNext(htsObserverAtomicList) { observer ->
                    observer.trySend(TemperatureMeasurement(celsius, fahrenheit))
                }
            }
            if (characteristic == HealthThermometer.TEMPERATURE_TYPE) {
                BleLogger.d(TAG, "TEMPERATURE_TYPE ${data.let { BleLogger.byteArrayToHex(it) }}")
            }
        }
    }

    override fun processServiceDataWritten(characteristic: UUID, status: Int) {
        BleLogger.d(TAG, "processServiceDataWritten TODO")
    }

    fun observeHtsNotifications(checkConnection: Boolean): Flow<TemperatureMeasurement> {
        return ChannelUtils.monitorNotifications(htsObserverAtomicList, txInterface, checkConnection)
            .onStart {
                d(TAG, "Start observing HTS")
                addCharacteristicNotification(HealthThermometer.TEMPERATURE_MEASUREMENT)
                txInterface.setCharacteristicNotify(
                    HealthThermometer.HTS_SERVICE,
                    HealthThermometer.TEMPERATURE_MEASUREMENT,
                    true
                )
            }
            .onCompletion {
                d(TAG, "Stop observing HTS")
                removeCharacteristicNotification(HealthThermometer.TEMPERATURE_MEASUREMENT)
                try {
                    txInterface.setCharacteristicNotify(
                        HealthThermometer.HTS_SERVICE,
                        HealthThermometer.TEMPERATURE_MEASUREMENT,
                        false
                    )
                } catch (e: Exception) {
                    // this may happen if connection is already closed, no need sent the exception to downstream
                    d(
                        TAG,
                        "HTS client is not able to set characteristic notify to false. Reason $e"
                    )
                }
            }
    }
}

File: sources/iOS/ios-communications/Sources/iOSCommunications/BleHtsClient.swift
import CoreBluetooth
import Foundation
import RxSwift

public class HealthThermometer {
    public static let HTS_SERVICE = CBUUID(string: "00001809-0000-1000-8000-00805f9b34fb")
    public static let TEMPERATURE_MEASUREMENT = CBUUID(string: "2A1C")
    public static let TEMPERATURE_TYPE = CBUUID(string: "2A1D")
}

public class BleHtsClient: BleGattClientBase {
    public struct TemperatureMeasurement {
        public let temperatureCelsius: Float
        public let temperatureFahrenheit: Float
    }

    public init(gattServiceTransmitter: BleAttributeTransportProtocol){
        super.init(serviceUuid: HealthThermometer.HTS_SERVICE, gattServiceTransmitter: gattServiceTransmitter)
        automaticEnableNotificationsOnConnect(chr: HealthThermometer.TEMPERATURE_MEASUREMENT, disableOnDisconnect:true)
    }
    
    static let TEMP_ACCURACY: Int = 100

    private let htsObserverAtomicList = AtomicList<RxObserver<TemperatureMeasurement>>()
    
    override public func disconnected() {
        super.disconnected()
        RxUtils.postErrorAndClearList(htsObserverAtomicList, error: BleGattException.gattDisconnected)
    }
    
    override public func processServiceData(_ chr: CBUUID, data: Data, err: Int ) {
        BleLogger.trace("processServiceData uuid=\(chr), err=\(err) len(data)=\(data.count)")
        
        if (err == 0) {
            if chr == HealthThermometer.TEMPERATURE_MEASUREMENT {
                BleLogger.trace_hex("TEMPERATURE_MEASUREMENT ", data: data)

                let flags = UInt8(data[0])
                let isFahrenheit = (flags & 0x01) != 0
                let exponent = Int8(bitPattern: data[4])

                let value = UInt32(data[0]) | (UInt32(data[1]) << 8) | (UInt32(data[2]) << 16) | (UInt32(data[3]) << 24)
                let mantissaMask: UInt32 = 0x007FFFFF
                let mantissa = (value & mantissaMask) >> 8

                let temperature = (Float(mantissa) * pow(10.0, Float(exponent)) * Float(BleHtsClient.TEMP_ACCURACY)).rounded() / Float(BleHtsClient.TEMP_ACCURACY)

                let celsius = !isFahrenheit ? temperature : (temperature - 32.0) * 5.0 / 9.0
                let fahrenheit = isFahrenheit ? temperature : temperature * 9.0 / 5.0 + 32.0
                
                RxUtils.emitNext(htsObserverAtomicList) { (observer) in
                    observer.obs.onNext(TemperatureMeasurement(temperatureCelsius: celsius, temperatureFahrenheit: fahrenheit))
                }
            }
            
            if chr == HealthThermometer.TEMPERATURE_TYPE {
                BleLogger.trace_hex("TEMPERATURE_TYPE ", data: data)
            }
        }
    }
    
    public func observeHtsNotifications(checkConnection: Bool) -> Observable<TemperatureMeasurement> {
         return RxUtils.monitor(htsObserverAtomicList, transport: self.gattServiceTransmitter, checkConnection: checkConnection)
    }
}

